### PR TITLE
Plans: descending price / description on hover abstest

### PR DESCRIPTION
**Pull request by artpi**
Introduces Abtests mentioned in #6997.

- Decreasing price order
- Description on hover

### [📺 Watch video](https://youtu.be/pu03RLBBzlw)

# NUX

zrzut ekranu 2016-08-15 o 11 58 32

# Plans

zrzut ekranu 2016-08-15 o 11 59 22

## Testing

1. Check out
2. Assign yourself to plansDescriptions: descendingPriceEagerDescription test group using env badge
3. Go to /plans. see that plans are sorted in descending price
4. Assign yourself ascendingPriceEagerDescription - see that descriptions open on hover
5. Test also ascendingPriceSubtleDescription ( control ) and descendingPriceSubtleDescription
4. Repeat all of that in nux

CC @rralian @lamosty @gwwar @retrofox 

Test live: https://calypso.live/?branch=new/plan-reorder-test

### CODE REVIEW

**Comment by lamosty** [comment]
Nice work! As far as code and functionality is concerned, it's all working fine :+1:.

For the abtest, I'm not sure we want it now etc, so let's wait for @rralian's word on this before merging. :)

**Comment by artpi** [comment]
Thanks for reviewing!
Yeah, I think we should wait - hence the 1.09 test start date :)

**Comment by gwwar** [comment]
👍 LGTM

**Comment by rralian** [comment]
> Yeah, I think we should wait - hence the 1.09 test start date :)

Yeah, agreed. We already have two tests running that are meant to impact plans. So I think waiting for the pricing test to complete is a very good idea.

**Comment by artpi** [comment]
As discussed in hangout with @rralian @retrofox @lamosty and @gwwar , I will:
- Rebase this from master to get @retrofox popover refactors
- Merge the test intended in https://github.com/Automattic/wp-calypso/pull/7465
- Create 4 variants:
    - ascending price / normal descriptions ( control, in master right now )
    - ascending price / descriptions on-row-hover ( test )
    - descending price / normal descriptions ( test )
    - descending price / descriptions on-row-hover ( test )

**Comment by artpi** [comment]
I updated the top comment with video and new testing instructions.
One caveat of interaction of new refactored popover and my changes is that on right column the popover displays right of row instead of right-of-icon.
I actually like this behavior.

I also created a commit that gets rid of that and I think its much worse:
- commit https://github.com/Automattic/wp-calypso/commit/9f73f669027fd08b9d1c9c7f99c55d5978cd0bce
- videeo: https://www.youtube.com/watch?v=1H5Noe91Ho0&feature=youtu.be

Course of action I intend to take is:

- Close https://github.com/Automattic/wp-calypso/pull/7465
- Close https://github.com/Automattic/wp-calypso/pull/7610
- Close https://github.com/Automattic/wp-calypso/pull/7608

Please advise.

File: client/my-sites/plan-features/item.jsx [hunk 1]
```diff
@@ -27,15 +28,25 @@ export default function PlanFeaturesItem( {
...
+	};
+	const hoverOnRow = (
+		abtest( 'plansDescriptions' ) === 'ascendingPriceEagerDescription' ||
+		abtest( 'plansDescriptions' ) === 'descendingPriceEagerDescription'
+	);
+
 	return (
-		<div className="plan-features__item">
+		<div className="plan-features__item"
+			{ ...( hoverOnRow && mouseEvents ) }
```
Comment (retrofox) [inline client/my-sites/plan-features/item.jsx]: Keep in mind that the target element (the `context` Popover property) is gotten from the function bound to the event.
In this case the function is `handleOnMouseEvent`, the event is `onMouseEnter` and the target is `<div className="plan-features__item">`.
For this reason the Popover isn't anchored to the `(i)` element.

image

Comment (artpi) [inline client/my-sites/plan-features/item.jsx]: Yes, I noticed and tried to changed that and decided its actually better.
> One caveat of interaction of new refactored popover and my changes is that on right column the popover displays right of row instead of right-of-icon. I actually like this behavior. I also created a commit that gets rid of that and I think its much worse:

- commit 9f73f66
- videeo: https://www.youtube.com/watch?v=1H5Noe91Ho0&feature=youtu.be

File: client/my-sites/plans-features-main/index.jsx [hunk 2]
```diff
@@ -83,15 +84,28 @@ class PlansFeaturesMain extends Component {
...
-			[
-				hideFreePlan ? null : PLAN_FREE,
-				isPersonalPlanEnabled ? PLAN_PERSONAL : null,
-				PLAN_PREMIUM,
-				PLAN_BUSINESS
-			],
-			value => !! value
-		);
+		let plans;
+		if ( abtest( 'plansDescriptions' ) === 'descendingPriceSubtleDescription' || abtest( 'plansDescriptions' ) === 'descendingPriceEagerDescription' ) {
```
Comment (retrofox) [inline client/my-sites/plans-features-main/index.jsx]: This line is too long triggering an eslint warning. I suggest:

```es6
		if (
			abtest( 'plansDescriptions' ) === 'descendingPriceSubtleDescription' ||
			abtest( 'plansDescriptions' ) === 'descendingPriceEagerDescription'
		) { 
			// ...
		}
```

Comment (artpi) [inline client/my-sites/plans-features-main/index.jsx]: Done!

File: client/my-sites/plan-features/item.jsx (hunk 1 above)
Comment (retrofox) [inline client/my-sites/plan-features/item.jsx]: oh. sorry. IMHO the Popover should be connected with `(i)` icon but let's wait to other opinions. Maybe ping to @mtias ?

Comment (gwwar) [inline client/my-sites/plan-features/item.jsx]: What if we target the `(i)` but choose a different default direction. For example top or below, so this doesn't cover the line text.

Comment (retrofox) [inline client/my-sites/plan-features/item.jsx]: I prefer that or also I could think in removing the `(i)` icon if the Poponer ins't connected direstly whit this.

File: client/my-sites/plan-features/item.jsx [hunk 3]
```diff
@@ -27,15 +28,25 @@ export default function PlanFeaturesItem( {
 		onMouseLeave( event.currentTarget, description );
 	};
 
+	const mouseEvents = {
+		onMouseEnter: handleOnMouseEvent,
+		onMouseLeave: handleOnMouseLeave
```
Comment (lamosty) [inline client/my-sites/plan-features/item.jsx]: Nit: If the event handler for `onMouseLeave` is called `handleOnMouseLeave`, wouldn't it be better if we called the event handler for `onMouseEnter`: `handleOnMouseEnter` instead of `handleOnMouseEvent`?

File: client/my-sites/plan-features/item.jsx (hunk 1 above)
Comment (lamosty) [inline client/my-sites/plan-features/item.jsx]: Okay, we need a definite resolution on this before merging. I think that while removing the (i) icon might be helpful to reduce the UI clutter ("look at all the small (i) icons on the plans page), it might require additional changes and delay the test.

I would say let's merge this already. :) But I'm waiting for @rralian to give you thumps up on this one, though.

File: client/my-sites/plan-features/item.jsx (hunk 3 above)
Comment (artpi) [inline client/my-sites/plan-features/item.jsx]: Yes, totally. Also I believe it's outside of the scope of this change (I dont want another rabbithole of refactoring popover)

Comment (lamosty) [inline client/my-sites/plan-features/item.jsx]: Oh, just noticed that `handleOnMouseEvent` is not part of this PR. Alrighty then. :)

**Comment by rralian** [comment]
> I would say let's merge this already. :) But I'm waiting for @rralian to give you thumps up on this one, though.

The handling of the popover in the "eager" variants on the last row isn't ideal.

> What if we target the (i) but choose a different default direction. For example top or below, so this doesn't cover the line text.

I agree this would be better. I can trigger something kinda like this at times.

better alignment image

but I think for the purposes of a quick and dirty test this tells us if eagerly showing details is helpful or distracting. My vote is to :shipit: as-is so we can get these tests running. And if it's simple to adjust the behavior of the last row, we can do so while the test is in place. If it's not simple, we can wait to see if the test gives us an indication that this is worth it.

File: client/lib/abtest/active-tests.js [hunk 4]
```diff
@@ -41,6 +41,17 @@ module.exports = {
 		},
 		defaultVariation: 'singlePurchaseFlow'
 	},
+	plansDescriptions: {
+		datestamp: '20160901',
```
Comment (rralian) [inline client/lib/abtest/active-tests.js]: Just remember to adjust the date prior to launch.

**Comment by folletto** [comment]
My suggestion there would be to have it attached to the `(i)` since it's the pattern we use everywhere, and should shop up above the box (so basically like [this screenshot](https://camo.githubusercontent.com/35d9168a42bfd44d47dad23c7e3dd510b5975786/68747470733a2f2f636c6f756475702e636f6d2f63624b5a7a7936326a32782b), but with the arrow on the `(i)`), so the full box is visible and the extra text appears. Above it because ideally people read from top to bottom, so they will then go to the next item below, which would be still visible.

Also, on mobile, above should work better than side.
